### PR TITLE
Added a CodecJson instance for Location

### PR DESCRIPTION
**Pull request by etorreborre**
+ exercised the corresponding law in the code.

### CODE REVIEW

File: notion-core/src/main/scala/com/ambiata/notion/core/Location.scala [hunk 1]
```diff
@@ -84,4 +86,27 @@ object Location {
...
+
+  def tagged[A](tag: String, c: HCursor, decoder: DecodeJson[A]): DecodeResult[A] =
+    (c --\ tag).hcursor.fold(DecodeResult.fail[A]("Invalid tagged type", c.history))(decoder.decode)
+
+  implicit def LocationCodecJson: CodecJson[Location] =
+    CodecJson.derived(LocationEncodeJson, LocationDecodeJson)
+
+
+  implicit def LocationEqual: Equal[Location] =
+    Equal.equal(_ == _)
```
Comment (markhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/Location.scala]: `Equal.equal(_ == _)` is just `Equal.equalA`.

File: notion-core/src/test/scala/com/ambiata/notion/core/LocationSpec.scala [hunk 2]
```diff
@@ -13,5 +16,9 @@ class LocationSpec extends Specification { def is = s2"""
    ${ fromUri("file:/hello/world") ==== LocalLocation("/hello/world").right }
    ${ fromUri("s3://hello/world") ==== S3Location("hello", "world").right }
 
+ A Location can be serialized/deserialized to/from Json $json
+
 """
+
+  def json = prop(implicitly[CodecJson[Location]].codecLaw.encodedecode _)
```
Comment (markhibberd) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationSpec.scala]: @charleso Was this encoded somewhere in ivory or somewhere else - can't remember?

File: notion-core/src/main/scala/com/ambiata/notion/core/Location.scala [hunk 3]
```diff
@@ -84,4 +86,27 @@ object Location {
...
+  implicit def LocationDecodeJson: DecodeJson[Location] =
+    DecodeJson(c =>
+      tagged("s3",    c, jdecode2L(S3Location.apply)("bucket", "key")).map(l => l:Location) |||
+      tagged("hdfs",  c, jdecode1L(HdfsLocation.apply)("path")).map(l => l:Location) |||
+      tagged("local", c, jdecode1L(LocalLocation.apply)("path")).map(l => l:Location))
+
+  def tagged[A](tag: String, c: HCursor, decoder: DecodeJson[A]): DecodeResult[A] =
+    (c --\ tag).hcursor.fold(DecodeResult.fail[A]("Invalid tagged type", c.history))(decoder.decode)
+
+  implicit def LocationCodecJson: CodecJson[Location] =
```
Comment (markhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/Location.scala]: This shouldn't exist, or shouldn't be implicit `CodecJson` isn't really a typeclass, it is just a kludge that allows encode and decode to be defined at once, so should either have a single CodecJson, or Encode and Decode but not both.

**Comment by markhibberd** [comment]
:shaved_ice: After implicit fix.

File: notion-core/src/test/scala/com/ambiata/notion/core/LocationSpec.scala (hunk 2 above)
Comment (charleso) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationSpec.scala]: > Was this encoded somewhere in ivory or somewhere else

[Yup](https://github.com/ambiata/ivory/blob/master/ivory-core/src/test/scala/com/ambiata/ivory/core/ArgonautProperties.scala)

Would be handy to have an argonaut-scalacheck. Worth trying to get in to the official project?

**Pull request merged**